File: frontend/src/lib/routePlanner.ts
const API_BASE_URL = import.meta.env.VITE_API_BASE_URL ?? '';
const ROUTE_PLANNER_FLAG = import.meta.env.VITE_ENABLE_ROUTE_PLANNER ?? 'false';

export type RouteCoordinate = [number, number];

export type RoutePlannerRoute = {
  id?: string;
  polyline: RouteCoordinate[];
  distance_km: number;
};

export type RoutePlannerVehicle = {
  battery_kwh: number;
  current_soc: number;
  target_arrival_soc: number;
  consumption_kwh_per_km: number;
  preferred_connector_types: string[];
  max_charging_kw: number;
};

export type RoutePlannerConstraints = {
  corridor_width_km?: number;
  max_results?: number;
};

export type RouteChargingPlanRequest = {
  route: RoutePlannerRoute;
  vehicle: RoutePlannerVehicle;
  constraints?: RoutePlannerConstraints;
  reference_time?: string;
};

export type RoutePlannerSummary = {
  distance_km: number;
  estimated_energy_kwh: number;
  start_soc: number;
  target_arrival_soc: number;
  minimum_required_soc: number;
  reachable_without_stop: boolean;
};

export type RoutePlannerRecommendation = {
  station_id: string;
  name: string;
  connector_type: string;
  max_kw: number;
  distance_from_route_km: number;
  route_distance_km: number;
  estimated_arrival_soc: number;
  score: number;
  reasons: string[];
};

export type RoutePlannerResponseMeta = {
  source: string;
  limitations: string[];
  snapshot_date?: string | null;
  freshness_label?: string | null;
};

export type RouteChargingPlanResponse = {
  route_id: string;
  summary: RoutePlannerSummary;
  recommendations: RoutePlannerRecommendation[];
  meta: RoutePlannerResponseMeta;
};

export function isRoutePlannerFlagEnabled(flagValue: string | undefined): boolean {
  return flagValue === 'true';
}

export function buildChargingPlanUrl(apiBaseUrl = API_BASE_URL): string {
  const baseUrl = apiBaseUrl.replace(/\/$/, '');
  return `${baseUrl}/api/routes/charging-plan`;
}

export async function fetchChargingPlan(
  request: RouteChargingPlanRequest,
  apiBaseUrl = API_BASE_URL,
): Promise<RouteChargingPlanResponse> {
  const response = await fetch(buildChargingPlanUrl(apiBaseUrl), {
    method: 'POST',
    headers: { 'Content-Type': 'application/json' },
    body: JSON.stringify(request),
  });

  if (!response.ok) {
    throw new Error(`Route planner failed: ${String(response.status)}`);
  }

  return response.json() as Promise<RouteChargingPlanResponse>;
}

export const routePlannerEnabled = isRoutePlannerFlagEnabled(ROUTE_PLANNER_FLAG);

File: backend/app/schemas/route_planner.py
from typing import Any

from pydantic import BaseModel, ConfigDict, Field


class RouteChargingPlanRequest(BaseModel):
    model_config = ConfigDict(extra="allow")

    route: Any = None
    vehicle: Any = None
    constraints: Any = None
    reference_time: Any = None

    def to_graph_payload(self) -> dict[str, Any]:
        return self.model_dump(exclude_none=True)


class RoutePlannerSummary(BaseModel):
    distance_km: float = Field(gt=0)
    estimated_energy_kwh: float = Field(ge=0)
    start_soc: float = Field(ge=0, le=1)
    target_arrival_soc: float = Field(ge=0, le=1)
    minimum_required_soc: float = Field(ge=0)
    reachable_without_stop: bool


class RoutePlannerRecommendation(BaseModel):
    station_id: str
    name: str
    connector_type: str
    max_kw: float = Field(gt=0)
    distance_from_route_km: float = Field(ge=0)
    route_distance_km: float = Field(ge=0)
    estimated_arrival_soc: float
    score: float
    reasons: list[str]


class RoutePlannerResponseMeta(BaseModel):
    source: str
    limitations: list[str]
    snapshot_date: str | None = None
    freshness_label: str | None = None


class RouteChargingPlanResponse(BaseModel):
    route_id: str
    summary: RoutePlannerSummary
    recommendations: list[RoutePlannerRecommendation]
    meta: RoutePlannerResponseMeta
